File: Combination Sum II/Combination Sum II.java
public class Solution {
    
    private void helper(List<List<Integer>> ret, List<Integer> sol, int start, int target, int[] num){
        if(target < 0)      return;
        else if(target == 0){
            ret.add(new ArrayList<Integer>(sol));
            return;
        }
        else{
            for(int i = start; i < num.length; ++i){
                sol.add(num[i]);
                target -= num[i];
                helper(ret, sol, i + 1, target, num);
                target += num[i];
                sol.remove(sol.size() - 1);
                while(i < num.length - 1 && num[i + 1] == num[i])       ++i;
            }
        }
    }
    
    public List<List<Integer>> combinationSum2(int[] num, int target) {
        
        Arrays.sort(num);
        List<Integer> sol = new ArrayList<Integer>();
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        helper(res, sol, 0, target, num);
        return res;
        
    }
}

File: Combination Sum II/Combination Sum II.cpp
class Solution {
public:

    void comb(std::vector<std::vector<int> >& ret, std::vector<int>& sol, std::vector<int>& num, int target, int index){
        if(target == 0){
            ret.push_back(sol);
            return;
        }
        else if(target < 0 || index == num.size()){
            return;
        }
        else{
            for(int i = index; i < num.size(); ++i){
                int getNum = num[i];
                target -= getNum;
                sol.push_back(getNum);
                comb(ret, sol, num, target, i + 1);
                sol.pop_back();
                target += getNum;
                while(i != num.size() - 1 && num[i + 1] == getNum){
                    ++i;
                }
            }
            return;
        }
    }

    vector<vector<int> > combinationSum2(vector<int> &num, int target) {
        std::vector<std::vector<int> > ret;
        std::vector<int> sol;
        std::sort(num.begin(), num.end());
        comb(ret, sol, num, target, 0);
        return ret;
    }
};
